File: Ruby/Irrgarten/dice.rb
class Dice
    @@MAX_USES = 5
    @@MAX_INTELLIGENCE = 10.0
    @@MAX_STRENGTH = 10.0
    @@RESURRECT_PROB = 0.3
    @@WEAPONS_REWARD = 2
    @@SHIELDS_REWARD = 3
    @@HEALTH_REWARD = 5
    @@MAX_ATTACKS = 3
    @@MAX_SHIELDS = 2

    @@generator = Random.new

    def self.random_pos(max)
        return @@generator.rand(max)
    end

    def self.who_starts(nplayers)
        return @@generator.rand(nplayers)
    end

    def self.random_intelligence()
        return @@generator.rand(@@MAX_INTELLIGENCE)
    end

    def self.random_strength()
        return @@generator.rand(@@MAX_STRENGTH)
    end

    def self.resurrect_player()
        return @@generator.rand() < @@RESURRECT_PROB
    end

    def self.weapon_reward()
        return @@generator.rand(@@WEAPONS_REWARD)
    end

    def self.shield_reward()
        return @@generator.rand(@@SHIELDS_REWARD)
    end

    def self.health_reward()
        return @@generator.rand(@@HEALTH_REWARD)
    end

    def self.weapon_power()
        return @@generator.rand(@@MAX_ATTACKS)
    end

    def self.shield_power()
        return @@generator.rand(@@MAX_SHIELDS)
    end

    def self.uses_left()
        return @@generator.rand(@@MAX_USES)
    end

    def self.intensity(competence)
        return @@generator.rand(competence)
    end

    def self.discard_element(uses_left)
        return @@generator.rand(1.0) < 1.0 - (uses_left/@@MAX_USES)
    end

    def self.next_step(preference, validMoves, intelligence)
        if(@@generator.rand(@@MAX_INTELLIGENCE) < intelligence)
            return preference
        else
            index = @@generator.rand(validMoves.size())
            return validMoves[index]
        end
    end
end

File: Java/irrgarten/Dice.java
package irrgarten;

import java.util.ArrayList;
import java.util.Random;

public class Dice {
    static private final int MAX_USES = 5;
    static private final float MAX_INTELLIGENCE = 10f;
    static private final float MAX_STRENGTH = 10f;
    static private final float RESURRECT_PROB = 0.3f;
    static private final int WEAPON_REWARD = 2;
    static private final int SHIELD_REWARD = 3;
    static private final int HEALTH_REWARD = 5;
    static private final int MAX_ATTACKS = 3;
    static private final int MAX_SHIELD = 2;

    static private final Random generator = new Random();

    static public int randomPos(int max) {
        return generator.nextInt(max);
    }

    static public int whoStarts(int nplayers) {
        return generator.nextInt(nplayers);
    }

    static public float randomIntelligence() {
        return generator.nextFloat() * MAX_INTELLIGENCE;
    }

    static public float randomStrength() {
        return generator.nextFloat() * MAX_STRENGTH;
    }

    static public boolean resurrectPlayer() {
        return generator.nextFloat() < RESURRECT_PROB;
    }

    static public int weaponsReward() {
        return generator.nextInt(WEAPON_REWARD);
    }

    static public int shieldsReward() {
        return generator.nextInt(SHIELD_REWARD);
    }

    static public int healthReward() {
        return generator.nextInt(HEALTH_REWARD);
    }

    static public float weaponPower() {
        return generator.nextFloat() * MAX_ATTACKS;
    }

    static public float shieldPower() {
        return generator.nextFloat() * MAX_SHIELD;
    }

    static public int usesLeft() {
        return generator.nextInt(MAX_USES);
    }

    static public float intensity(float competence) {
        return generator.nextFloat() * competence;
    }

    static public boolean discardElement(int usesLeft) {
        return generator.nextFloat() < 1.0f - (usesLeft / MAX_USES);
    }

    static public Directions nextStep(Directions preferente, ArrayList<Directions> validMoves, float intelligence) {
        if (generator.nextFloat() * MAX_INTELLIGENCE < intelligence) {
            return preferente;
        } else {
            int index = generator.nextInt(validMoves.size());
            return validMoves.get(index);
        }
    }
}
